### PR TITLE
Turn kafka into a plugin

### PR DESCRIPTION
**Pull request by fireworkmarks**

### CODE REVIEW

File: messagebus/message_sender.go [hunk 1]
```diff
@@ -20,33 +16,18 @@ type MessageSender interface {
...
-		if !ok {
-			return nil, errors.New("msg_bus config is invalidate.")
-		}
-		MsgSender, err = builder.Create(helper.CONFIG.MsgBus)
-		if err != nil || nil == MsgSender {
-			return nil, errors.New(fmt.Sprintf("failed to create message_sender with err: %v", err))
-		}
-
-		atomic.StoreUint32(&initialized, 1)
+	plugins := mods.InitialPlugins()
```
Comment (cxt90730) [inline messagebus/message_sender.go]: Initial one time

File: messagebus/message_sender.go [hunk 2]
```diff
@@ -1,52 +1,31 @@
...
-		MsgSender, err = builder.Create(helper.CONFIG.MsgBus)
-		if err != nil || nil == MsgSender {
-			return nil, errors.New(fmt.Sprintf("failed to create message_sender with err: %v", err))
-		}
-
-		atomic.StoreUint32(&initialized, 1)
+func GetMessageSender(plugins map[string]*mods.YigPlugin) (MessageSender, error) {
+	name := "kafka"
+	p := plugins[name]
+	c, err := p.Create(helper.CONFIG.Plugins[name].Args)
```
Comment (cannium) [inline messagebus/message_sender.go]: No longer a singleton?

File: main.go [hunk 3]
```diff
@@ -75,9 +78,6 @@ func main() {
 		helper.Logger.Info("Succeed to create message bus sender.")
 	}
 
-	// Read all *.so from plugins directory, and fill the variable allPlugins
-	allPluginMap := mods.InitialPlugins()
-
 	iam.InitializeIamClient(allPluginMap)
```
Comment (cxt90730) [inline main.go]: Whether the `iam.InitializeIamClient(allPluginMap)` is a repeated work?

Comment (fireworkmarks) [reply main.go]: emmm, probably not, it loads according to the name it chooses, and it is initialized only once

File: messagebus/message_sender.go (hunk 2 above)
Comment (fireworkmarks) [reply messagebus/message_sender.go]: I probably could not understand

File: plugins/kafka_plugin.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	}
+	kf.producer.Flush(timeout)
+	return nil
+}
+
+func (kf *Kafka) Close() {
+	if nil == kf.producer {
+		return
+	}
+	kf.producer.Flush(300000)
```
Comment (cxt90730) [inline plugins/kafka_plugin.go]: what's the flush() mean?

File: collector.go [hunk 5]
```diff
@@ -81,65 +85,76 @@ func (c *Metrics) GenerateBucketUsageData() (GaugeMetricData map[string]UsageDat
...
 			return
 		}
-		if data.storageClass != "" {
-			GaugeMetricData[bucket.Name] = UsageDataWithBucket{data.value, bucket.OwnerId, data.storageClass}
+		for _, data := range datas {
+			if len(GaugeMetricData[bucket.Name]) == 0 {
+				GaugeMetricData[bucket.Name] = []UsageDataWithBucket{{data.value, bucket.OwnerId, data.storageClass}}
+			} else {
+				GaugeMetricData[bucket.Name] = append(GaugeMetricData[bucket.Name], UsageDataWithBucket{data.value, bucket.OwnerId, data.storageClass})
+			}
```
Comment (cannium) [inline collector.go]: It's OK to append to a 0-length slice. e.g.
```go
	var l []int
	l = nil
	l = append(l, 4)
	l = append(l, 5)
```

File: mods/yig-plugins.go [hunk 6]
```diff
@@ -24,6 +24,7 @@ const EXPORTED_PLUGIN = "Exported"
 
 const (
 	IAM_PLUGIN = iota //IamClient interface
+	KAFKA_PLUGIN
```
Comment (cxt90730) [inline mods/yig-plugins.go]: I prefer to modify name to 'MQ_PLUGIN', and mq type is 'kafka'

File: plugins/dummy_msgbus_plugin.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	return interface{}(msgbus), nil
+}
+
+type dummyMsgBus struct {
+	Url   string
+	Topic string
+}
+
+func (mb *dummyMsgBus) Start() error {
+	helper.Logger.Println("Start message bus succeed! url is:", mb.Url, "topic is:", mb.Topic)
```
Comment (cxt90730) [inline plugins/dummy_msgbus_plugin.go]: fmt

**Pull request merged**